### ISSUE TITLE
Gherkin: documentation for semantics of `Rule` keyword

### ISSUE THREAD

**Issue by bbqsrc** [cucumber/cucumber#453]
Other than #416, I haven't been able to find any documentation about the Rule keyword and its expected children elements, nor what "Gherkin 6" is to consist of. Has this yet been documented anywhere?

**Comment by aslakhellesoy** [comment]
@bbqsrc Gherkin 6 isn't released yet, nor is it integrated into any Cucumber implementations yet. It's my primary focus at the moment.

You will find more background in #250, which is the issue that prompted the addition. The changes in Gherkin 6 will be finalised when we make a new release. I might as well start here though:

## Rule keyword

The new optional `Rule` keyword is optional. Older Gherkin documents without `Rule` keywords are still valid.

The `Rule` keyword is simply a grouping construct. It can be used to group `Scenario`s that are related to the same `Rule`. This makes it possible to represent an [example map](http://examplemapping.com) in a Gherkin document. Some people may want to replace `Scenario` with `Example` (a new synonym to `Scenario`) to make Gherkin documents even more closely aligned with example mapping.

## Scenario Outline = Scenario

It is now possible to place `Examples` tables under a regular `Scenario`. When you do, the scenario is executed several times, replacing `<>` tokens with values from each `Examples` row. The `Scenario Outline` keyword is now just a synonym for `Scenario`.

## UTF-8

Previous versions of Gherkin made it possible to specify an alternate encoding for `.feature` files using a header such as `# encoding: ISO-8859-1`. This is no longer possible. All Gherkin documents must be UTF-8 encoded.

**Comment by bbqsrc** [comment]
Thanks for that information. It seems so far I have been on track with how Rule is meant to behave, and Scenario Outline = Outline, in the [Rust implementation of Cucumber](https://github.com/bbqsrc/cucumber-rust) and [Gherkin](https://github.com/bbqsrc/cucumber-rust) I've been working on.

I will do my best to follow Gherkin 6 as it continues to emerge. 😄

**Comment by aslakhellesoy** [comment]
Oooh a Cucumber-Rust would be amazing! 

I recommend you take a look at the [Cucumber roadmap](https://docs.google.com/document/d/12Y3Ut-SVSGdw9L_aW6NGQY9JaCcn3z0alY4d7kzOndU).

All the official Cucumber implementations will use [gherkin-go](https://github.com/cucumber/cucumber/tree/master/gherkin/go) (over protobuf). A full Cucumber implementation has several components (gherkin, [cucumber-expressions](https://github.com/cucumber/cucumber/tree/master/cucumber-expressions), [tag-expressions](https://github.com/cucumber/cucumber/tree/master/tag-expressions), [datatable](https://github.com/cucumber/cucumber/tree/master/datatable), formatters etc). 

To reduce the amount of code that needs to be implemented (and maintained) we're working on a [cucumber-engine]() in Go. This will make it easy to write a new Cucumber implementation in any language.

Would you consider joining us on this roadmap?

**Comment by aslakhellesoy** [comment]
Ok, sounds good. The NDJSON representation of messages is only used for gherkin's approval tests. Inter-process communication (e.g. Java requesting Gherkin ASTs and pickles from `gherkin-go`) will use protobuf.

As for the Gherkin grammar spec - it's in [gherkin.berp](https://github.com/cucumber/cucumber/blob/master/gherkin/gherkin.berp). The parser generator is [berp](), which we built to support several languages. Now that we're only targeting Go, we could use a standard parser generator (or even hand-write a recursive descent parser - it's a simple grammar), but we don't see a need to - it's working.

**Comment by bbqsrc** [comment]
Is there a more appropriate venue to discuss these kinds of things or is
GitHub the place?


On 12 Aug. 2018 at 18:34, <Aslak Hellesøy <notifications@github.com>>
wrote:

Ok, sounds good. The NDJSON representation of messages is only used for
gherkin's approval tests. Inter-process communication (e.g. Java requesting
Gherkin ASTs and pickles from gherkin-go) will use protobuf.

As for the Gherkin grammar spec - it's in gherkin.berp
<https://github.com/cucumber/cucumber/blob/master/gherkin/gherkin.berp>.
The parser generator is berp, which we built to support several languages.
Now that we're only targeting Go, we could use a standard parser generator
(or even hand-write a recursive descent parser - it's a simple grammar),
but we don't see a need to - it's working.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/cucumber/cucumber/issues/453#issuecomment-412354800>,
or mute the thread

**Comment by aslakhellesoy** [comment]
> Is there a more appropriate venue to discuss these kinds of things or is GitHub the place?

Yes, the `#committers` channel in Slack. See https://cucumber.io/support for details about how to join.

**Comment by aurelien-reeves** [comment]
@aslakhellesoy @bbqsrc what about that issue? It seems outdated to me:

- gherkin-go has not been used as the only gherkin implementation, and wont be
- protobuf is not a thing anymore
- the rule keyword is now well supported

@bbqsrc do you still need it?

PS: last time I looked at cucumber-rust, it looked pretty nice! Well done! 👏
